JDK-8287284: C2: loop optimization performs split_thru_phi infinitely many times

tobiasholenstein opened the pull request:
### Node* phi = PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) 
The function `PhaseIdealLoop::split_thru_phi` splits Node `n` through a merge point (phi) if there is enough "win". First a clone x<sub>i</sub> of `n` is created for each input of the merge point. The general idea is that we want to replace enough of the x<sub>i</sub> clones with existing nodes or constants. The `policy` indicates how many "wins" are needed for success. A "win" is defined as one of the following:
- Identity(x<sub>i</sub>) has singleton type (i.e. a constant)
- Identity(x<sub>i</sub>) returns an already existing node
- x<sub>i</sub> commons up with an existing node (i.e. hash_find(x<sub>i</sub> != null)

_Example_
Looking at the phi (119 Phi) of `n` (135 LoadI) a clone of `n` is created for each input to the phi: 
- x<sub>1</sub> (140 LoadI) for the input [1] (127 Phi)
- x<sub>2</sub> (141 LoadI) for the input [2] (90 StoreI)

If there are enough "wins" (win $\geq$ policy) the wins are applied and a new merge point `phi` (139 Phi) is returned. 
`n` (135 LoadI) is replaced with `phi` (139 Phi) outside of `PhaseIdealLoop::split_thru_phi` if successful.

Example with **n =** `135 LoadI`, **region =** `129 Loop`, **policy =** `1` returns **phi =** `139 Phi`
<img width="286" alt="split_phi" src="https://github.com/openjdk/jdk/assets/71546117/4a1bcd8d-bbfd-4267-9d2b-5ba4e057df72">
Success because 1 "win" occurs: `88 AddI` = Identity(`141 LoadI`)
<img width="186" alt="88_identity" src="https://github.com/openjdk/jdk/assets/71546117/af3cb37a-c575-445a-b8d4-967013dd3fb0">



### run() has infinite loop of split_thru_phi
```java
// java -Xbatch -XX:-PartialPeelLoop -XX:CompileCommand=compileonly,Test::run Test.java

public class Test {

    public static int cnt = 1;

    public static void run() {
        int j = 0;
        do {
            j = cnt;
            for (int k = 0; k < 20000; k++) {
                cnt += 2;
            }

        } while (++j < 10);
    }

    public static void main(String[] args) {
        for (int i = 0; i < 10; i++ ) {
            run();
        }
    }
}
```
`Compile::optimize_loops` ends up in an endless loop when optimizing run().
In each `PhaseIdealLoop iteration` a `LoadI` (e.g. 135) is split through a `Phi` (e.g. 119) and the newly created `LoadI` (e.g. 140) is again split through a `Phi` (e.g. 127) creating an endless repeating pattern of `LoadI` (e.g. 144) always connected to the circular dependent `Phis` (119, 127). This results in an every growing chain of new `Phis` (142, 139, 133). The process is only stopped because `LoopOptsCount` is reached after 43 iterations. The long chain of `Phis` stays throughout code generation.
<img width="596" alt="endless" src="https://github.com/openjdk/jdk/assets/71546117/5d0353a2-3212-403d-9f78-5bf60ac8df4e">

Lets look at the steps of `split_thru_phi` in more detail:

1. The graph of `run()` before the first `PhaseIdealLoop iterations` looks like the following:
![1 - init graph - split 135](https://github.com/openjdk/jdk/assets/71546117/b8663eb6-e10d-4450-9a48-e366c4b47f6e)

2. Running `build_and_optimize()`&rarr;`split_if_with_blocks()`&rarr;`split_if_with_blocks_pre(use)`&rarr;`split_thru_phi(Node* n, Node* region, int policy)` with **n =** `135 LoadI`, **region =** `139 Loop`. 
- `135 LoadI` is split through `119 Phi` if it results in a "win". 
- As a first step a clone of `135 LoadI` is created for each input of `119 Phi`:
   - clone `140 LoadI` for input `127 Phi`
   - `141 LoadI` for input `90 StoreI`
![2 - clone 140 141](https://github.com/openjdk/jdk/assets/71546117/e024e264-0b51-45da-9027-385d9750a8a3)

3. In this case we have a "win" because there exists an identity of `141 LoadI`:
- `88 AddI` = identity(`141 LoadI`)
![3 - 88=identity(141)](https://github.com/openjdk/jdk/assets/71546117/bb8a2e3e-1ccb-4ea7-b0fa-812436bd4cec)

4. As a final step of `split_thru_phi` (and only because we have a "win") `135 LoadI` is replaced with a new `139 Phi` that merges the clone `140 LoadI` and `88 AddI`, the identity of `141 LoadI` which gets cleaned up:
![4 - replace(135 with 139)](https://github.com/openjdk/jdk/assets/71546117/b861967e-9a2d-4056-bf9d-611377800a62)

5. Next `split_thru_phi(Node* n, Node* region, int policy)` is called with the previously created clone **n =** `140 LoadI`.
- `140 LoadI` is split through `127 Phi`
- Clone `143 LoadI` is created for `7 Parm`, the first input to `127 Phi`
![5 - split 140 and clone 143](https://github.com/openjdk/jdk/assets/71546117/ebbe989c-ffb9-4ad4-8f79-c1cb321c886a)

6. We have again a "win" for the clone `143 LoadI` because:
- `24 Loadl` = hash(`143 LoadI`)
![6 - 24=hash(143)](https://github.com/openjdk/jdk/assets/71546117/7bb9587e-8b28-4788-8a2f-cc1da97c2f14)

7. Next, clone `144 LoadI` is created for `119 Phi`, the second input to `127 Phi`. And finally, `140 LoadI` is replaced with a new `142 Phi` 
![7 - replace(140 with 142)](https://github.com/openjdk/jdk/assets/71546117/72f0fc45-5898-42d8-a506-0ca08fd1c068)
If `split_thru_phi` was called on a LoadI, region is a Loop and we have a "win" , we do `set_major_progress()` (at the end of `PhaseIdealLoop::split_if_with_blocks_pre`). Therefore we do another `PhaseIdealLoop iteration`.
https://github.com/openjdk/jdk/blob/903b9e8dd966fbb61222c59048b752ed8b42b608/src/hotspot/share/opto/loopopts.cpp#L1171-L1174

8. In the second `PhaseIdealLoop iteration` we call `split_thru_phi(Node* n, Node* region, int policy)` with **n =** `144 LoadI`. This is the same as in the first iteration: `44 LoadI` has the same inputs as `135 LoadI` in the first iteration. We find the exact same wins again because the involved nodes have not changed in the meantime. We repeat the work of the first iteration and `set major_progress` to true. For each following loop opts round, we repeat the same steps again and keep setting `major_progress` to true until we reach the maximum number of loop opts iterations.

=> **We are stuck in a loop!**
![8 - split 144](https://github.com/openjdk/jdk/assets/71546117/ba340d10-5031-4b30-b97b-63793fc65118)

9. After another call to `split_thru_phi` we have a chain of 4 phis that grows until `PhaseIdealLoop iterations` runs out of iterations.
![9 - chain of phis](https://github.com/openjdk/jdk/assets/71546117/c282b4d7-57af-4fa4-a90d-6d69afba52ed)

# Proposed Fix

The problem is that `split_thru_phi` finds a win when in reality it is not a win: it does not make much sense to move the LoadI back into the inner loop - Because that means adding extra work to the most common path. The cost function in `split_thru_phi` is now adjusted to exclude such false wins. This automatically also prevents the infinite loop.


---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
 
### Issue
 * [JDK-8287284](https://bugs.openjdk.org/browse/JDK-8287284): C2: loop optimization performs split_thru_phi infinitely many times (**Bug** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15536/head:pull/15536` \
`$ git checkout pull/15536`

Update a local copy of the PR: \
`$ git checkout pull/15536` \
`$ git pull https://git.openjdk.org/jdk.git pull/15536/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15536`

View PR using the GUI difftool: \
`$ git pr show -t 15536`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15536.diff">https://git.openjdk.org/jdk/pull/15536.diff</a>

</details>

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287284](https://bugs.openjdk.org/browse/JDK-8287284): C2: loop optimization performs split_thru_phi infinitely many times (**Bug** - P4)


### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**) ⚠️ Review applies to [07cfafc7](https://git.openjdk.org/jdk/pull/15536/files/07cfafc7c8d07d352442295f56100c7781136277)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [8b7d5ce6](https://git.openjdk.org/jdk/pull/15536/files/8b7d5ce6b988a891f238603e6105bc85f43d7d73)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15536/head:pull/15536` \
`$ git checkout pull/15536`

Update a local copy of the PR: \
`$ git checkout pull/15536` \
`$ git pull https://git.openjdk.org/jdk.git pull/15536/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15536`

View PR using the GUI difftool: \
`$ git pr show -t 15536`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15536.diff">https://git.openjdk.org/jdk/pull/15536.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15536#issuecomment-1774958874)

TobiHartmann commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -218,6 +229,23 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
+bool PhaseIdealLoop::has_moved_to_inner_loop(Node* n, Node* region, Node* x) {
+  assert(region->is_Loop(), "region should be a loop");
+  IdealLoopTree* n_loop_tree = get_loop(region);
+  for (uint j = 1; j < n->req(); j++) {
```
Why do you need this loop at all? Wouldn't it be sufficient to just check `x_loop_tree->is_member(n_loop_tree)`?

tobiasholenstein replied:
you are right! I changed it

TobiHartmann reviewed (approved):
That looks good to me.

rwestrel commented:
`moved_to_inner_loop()` returns true if the clone may end up in an inner loop but it could return true and the clone be in the outer loop: `moved_to_inner_loop()` tests the early control but the clone will be assigned control that is as late as possible as long as it doesn't cause a node to move in a loop. In short that test is not exact. I would adjust the comment to reflect that.
I also think this happens because the node that's cloned is a `Load` and anti dependence analysis is conservative and pushes it in the inner loop so we could apply the `moved_to_inner_loop()` test only to loads. Not sure if it's worth doing though.

eme64 commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -218,6 +228,17 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
 }
 
+// Test whether node 'x' can move into an inner loop relative to node 'n'.
+// Note: The test is not exact. Returns true if the clone COULD end up in an inner loop,
```
```suggestion
// Note: The test is not exact. Returns true if the 'x' COULD end up in an inner loop,
```

eme64 commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -218,6 +228,17 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
 
+// Test whether node 'x' can move into an inner loop relative to node 'n'.
+// Note: The test is not exact. Returns true if the clone COULD end up in an inner loop,
+// BUT it can also return true and the clone is in the outer loop
```
```suggestion
// BUT it can also return true and the 'x' is in the outer loop
```

eme64 commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -218,6 +228,17 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
+// Note: The test is not exact. Returns true if the clone COULD end up in an inner loop,
+// BUT it can also return true and the clone is in the outer loop
+bool PhaseIdealLoop::can_move_to_inner_loop(Node* n, Node* region, Node* x) {
+  assert(region->is_Loop(), "region should be a loop");
```
make argument a `LoopNode`, then you can drop the assert.

eme64 commented:
> `moved_to_inner_loop()` returns true if the clone may end up in an inner loop but it could return true and the clone be in the outer loop: `moved_to_inner_loop()` tests the early control but the clone will be assigned control that is as late as possible as long as it doesn't cause a node to move in a loop. In short that test is not exact. I would adjust the comment to reflect that. I also think this happens because the node that's cloned is a `Load` and anti dependence analysis is conservative and pushes it in the inner loop so we could apply the `moved_to_inner_loop()` test only to loads. Not sure if it's worth doing though.

So you think this problem only exists for Loads? Can we not have a similar split_thru_phi problem with another node?

As you say: early control is not sufficient, as it could still have all uses outside the inner loop and would then be scheduled after the inner loop (ending up in the outer loop). I guess we could just check for the late control then?

rwestrel commented:
> So you think this problem only exists for Loads? Can we not have a similar split_thru_phi problem with another node?

The exit of the inner loop has to be some control split with a projection that's in the outer loop. The cloned node should be assigned that projection as control (in the outer loop) unless there's something pushing it in the inner loop which I assume in this case is a store. Not sure how that would happen with other nodes.

> As you say: early control is not sufficient, as it could still have all uses outside the inner loop and would then be scheduled after the inner loop (ending up in the outer loop). I guess we could just check for the late control then?

Yes, late control would be better but it's unclear to me how straightforward it is and given this is a corner case, a simple solution is likely good enough.

tobiasholenstein commented:
> > So you think this problem only exists for Loads? Can we not have a similar split_thru_phi problem with another node?
> 
> The exit of the inner loop has to be some control split with a projection that's in the outer loop. The cloned node should be assigned that projection as control (in the outer loop) unless there's something pushing it in the inner loop which I assume in this case is a store. Not sure how that would happen with other nodes.
> 
> > As you say: early control is not sufficient, as it could still have all uses outside the inner loop and would then be scheduled after the inner loop (ending up in the outer loop). I guess we could just check for the late control then?
> 
> Yes, late control would be better but it's unclear to me how straightforward it is and given this is a corner case, a simple solution is likely good enough.

We do not always have late control. So I lets stay with early control to keep the fix simple. 
Since testing with early control does not give an exact result, and the infinite loop only occurs for Loads, I suggest we apply the fix only for Loads. 

I updated the code and comments.

chhagedorn commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -152,9 +152,19 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
+      _igvn.remove_dead_node(the_clone);
+    } else if (region->is_Loop() && i == LoopNode::LoopBackControl &&
+               n->is_Load() && can_move_to_inner_loop(n, region->as_Loop(), x)) {
+      // it is not a win if 'x' moved from an outer to an inner loop
```
Maybe you want to add a comment here why we only want to do it for loads.

tobiasholenstein replied:
done

chhagedorn commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -152,9 +152,19 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
+
     phi->set_req( i, x );
+
+    if (the_clone == nullptr) continue;
```
Nit: add braces
```suggestion
    if (the_clone == nullptr) {
      continue;
    }
```

chhagedorn commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1735,6 +1735,8 @@ class PhaseIdealLoop : public PhaseTransform {
 
   void update_addp_chain_base(Node* x, Node* old_base, Node* new_base);
+
+  bool can_move_to_inner_loop(Node* n, LoopNode* region, Node* x);
```
Should match the signature in the cpp file:
```suggestion
  bool can_move_to_inner_loop(Node* n, LoopNode* n_loop, Node* x);
```

chhagedorn reviewed (approved):
Nice summary! I agree to go with a simple solution .

I have a few minor comments, otherwise, looks good!

tobiasholenstein commented:
I am ready to integrate. Do you agree @rwestrel ?

rwestrel reviewed (approved):
Looks good to me.

rwestrel commented:
> I am ready to integrate. Do you agree @rwestrel ?

What about a test case?

chhagedorn commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -152,9 +152,22 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
+    } else if (region->is_Loop() && i == LoopNode::LoopBackControl &&
+               n->is_Load() && can_move_to_inner_loop(n, region->as_Loop(), x)) {
+      // it is not a win if 'x' moved from an outer to an inner loop
+      // this edge case can only happend for Load nodes
```
```suggestion
      // this edge case can only happen for Load nodes
```

tobiasholenstein commented:
> > I am ready to integrate. Do you agree @rwestrel ?
> 
> What about a test case?

right! I added one now

rwestrel commented:
> > > I am ready to integrate. Do you agree @rwestrel ?
> > 
> > 
> > What about a test case?
> 
> right! I added one now

Does that test case cause a crash with the current jdk code? We used to have an assert on the number of rounds of loop opts but we don't have it anymore AFAICT.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/TestSplitThruPhiInfinitely.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @run main/othervm -Xbatch -XX:-PartialPeelLoop -XX:CompileCommand=compileonly,compiler.c2.loopopts.TestSplitThruPhiInfinitely::run* compiler.c2.loopopts.TestSplitThruPhiInfinitely
+
+
```
You should have a `*` on every line.

tobiasholenstein replied:
resolved now

chhagedorn commented:
> > > > I am ready to integrate. Do you agree @rwestrel ?
> > > 
> > > 
> > > What about a test case?
> > 
> > 
> > right! I added one now
> 
> Does that test case cause a crash with the current jdk code? We used to have an assert on the number of rounds of loop opts but we don't have it anymore AFAICT.

If not, you could think about adding an IR test that checks that we are not adding too many phis/loads. It's not perfect but at least something.

rwestrel commented:
> If not, you could think about adding an IR test that checks that we are not adding too many phis/loads. It's not perfect but at least something.

Or maybe parsing the output of `TraceLoopOpts`.

tobiasholenstein commented:
> If not, you could think about adding an IR test that checks that we are not adding too many phis/loads. It's not perfect but at least something.

I added an IR test on number of PHI's now. The test passes with the fix and fails without the fix.

tobiasholenstein commented:
> > > > I am ready to integrate. Do you agree @rwestrel ?
> > > 
> > > 
> > > What about a test case?
> > 
> > 
> > right! I added one now
> 
> Does that test case cause a crash with the current jdk code? We used to have an assert on the number of rounds of loop opts but we don't have it anymore AFAICT.

You are right. I added an IR test-case now that fails without the fix. I plan to reintroduce that assert in another RFE

rwestrel commented:
> You are right. I added an IR test-case now that fails without the fix. I plan to reintroduce that assert in another RFE

Thanks for adding the test case. That looks good to me.

chhagedorn reviewed (approved):
Thanks for adding a test!

tobiasholenstein commented:
Thanks @rwestrel , @eme64 , @eme64 and @TobiHartmann for the reviews and discussions!

/integrate
